File: app/src/main/java/apincer/android/mmate/ui/viewmodel/TagsViewModel.java
package apincer.android.mmate.ui.viewmodel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.function.Function;

import javax.inject.Inject;

import apincer.music.core.database.MusicTag;
import apincer.music.core.repository.TagRepository;
import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class TagsViewModel extends ViewModel {

    public static final String MULTIPLE_ALBUMS = "[Multiple Albums]";
    public static final String MULTIPLE_ARTISTS = "[Multiple Artists]";
    public static final String MULTIPLE_ALBUM_ARTISTS = "[Multiple Album Artists]";
    public static final String MULTIPLE_EMPTY = "";

    // --- Core Data ---
    private final MutableLiveData<List<MusicTag>> _editItems = new MutableLiveData<>(Collections.emptyList());
    public final LiveData<List<MusicTag>> editItems = _editItems;

    private final MutableLiveData<MusicTag> _displayTag = new MutableLiveData<>();
    public final LiveData<MusicTag> displayTag = _displayTag;

    private final TagRepository repos;

    @Inject
    public TagsViewModel(TagRepository repos) {
        this.repos = repos;
    }

    public void processAudioTagEditEvent(List<MusicTag> items) {
        //_searchCriteria.postValue(criteria);
        _editItems.postValue(items); // Post a copy
        rebuildDisplayTag(items);
    }

    public void updateWithPlayingSong(MusicTag playingSong) {
        if (playingSong == null) return;
        // This is a simplified version of your updatePreview logic
        // You might need more sophisticated logic to decide if the current editItems
        // should be completely replaced or if this is just for preview.
        repos.load(playingSong); // Assuming this is synchronous or you handle async
        List<MusicTag> singleItemList = new ArrayList<>();
        singleItemList.add(playingSong.copy()); // Work with a copy
        _editItems.postValue(singleItemList);
        rebuildDisplayTag(singleItemList);
    }

    private void rebuildDisplayTag(List<MusicTag> currentItems) {
        if (currentItems == null || currentItems.isEmpty()) {
            _displayTag.postValue(null);
            return;
        }
        // This is your logic from TagsActivity.buildDisplayTag()
        // For simplicity, I'm just taking the first item. Adapt as needed.
        MusicTag newDisplayTag = currentItems.get(0).copy(); // Make a copy for display
        if (currentItems.size() > 1) {
            // Your logic to create a "common" or summary tag
            newDisplayTag.setTitle("["+currentItems.size() + " songs selected]");
            // Use the generic helper for each field
            newDisplayTag.setAlbum(getCommonStringValue(currentItems, MusicTag::getAlbum, MULTIPLE_ALBUMS, MULTIPLE_EMPTY));
            newDisplayTag.setArtist(getCommonStringValue(currentItems, MusicTag::getArtist, MULTIPLE_ARTISTS, MULTIPLE_EMPTY));
            newDisplayTag.setAlbumArtist(getCommonStringValue(currentItems, MusicTag::getAlbumArtist, MULTIPLE_ALBUM_ARTISTS, MULTIPLE_EMPTY));
            newDisplayTag.setGenre(getCommonStringValue(currentItems, MusicTag::getGenre, MULTIPLE_EMPTY, MULTIPLE_EMPTY));
            newDisplayTag.setGrouping(getCommonStringValue(currentItems, MusicTag::getGrouping, MULTIPLE_EMPTY, MULTIPLE_EMPTY));
            newDisplayTag.setTrack(getCommonStringValue(currentItems, MusicTag::getTrack, MULTIPLE_EMPTY, MULTIPLE_EMPTY));
            newDisplayTag.setYear(getCommonStringValue(currentItems, MusicTag::getYear, MULTIPLE_EMPTY, MULTIPLE_EMPTY));
            newDisplayTag.setDisc(getCommonStringValue(currentItems, MusicTag::getDisc, MULTIPLE_EMPTY, MULTIPLE_EMPTY));
           // newDisplayTag.setMediaType(getCommonStringValue(currentItems, MusicTag::getMediaType, MULTIPLE_EMPTY, MULTIPLE_EMPTY));
            newDisplayTag.setPublisher(getCommonStringValue(currentItems, MusicTag::getPublisher, MULTIPLE_EMPTY, MULTIPLE_EMPTY));
            newDisplayTag.setQualityRating(getCommonStringValue(currentItems, MusicTag::getQualityRating, MULTIPLE_EMPTY, MULTIPLE_EMPTY));
        }
        _displayTag.postValue(newDisplayTag);
    }

    /**
     * Generic method to find a common string value from a list of MusicTags for a given field.
     *
     * @param tags List of MusicTag objects.
     * @param valueExtractor A function that takes a MusicTag and returns the String value of the desired field.
     * @param multipleValueString The string to return if multiple distinct non-null values are found, or if mixing nulls and non-nulls.
     * @param defaultValueForNullCommon If all non-null tags have a null value for the field, what to return (e.g., "", or null).
     * @return The common string value, multipleValueString, or defaultValueForNullCommon.
     */
    public static String getCommonStringValue(List<MusicTag> tags,
                                              Function<MusicTag, String> valueExtractor,
                                              String multipleValueString,
                                              String defaultValueForNullCommon) {
        if (tags == null || tags.isEmpty()) {
            return multipleValueString; // Or some other default for empty list
        }

        List<String> values = tags.stream()
                .filter(Objects::nonNull) // Ignore null MusicTag objects
                .map(valueExtractor)      // Extract the specific field's value (can be null)
                .toList();

        if (values.isEmpty()) {
            // All MusicTag objects in the original list were null
            return multipleValueString; // Or appropriate default
        }

        long distinctCount = values.stream().distinct().limit(2).count();

        if (distinctCount > 1) {
            return multipleValueString;
        } else { // distinctCount is 1 (or 0 if values was empty, handled above)
            String commonValue = values.get(0); // This is the common value (could be null)
            return commonValue != null ? commonValue : defaultValueForNullCommon;
        }
    }

    public void refreshDisplayTag() {
        List<MusicTag> items = _editItems.getValue();
        List<MusicTag> updatedItems = new ArrayList<>(); // Or refetch
        //should reload music tags
        items.forEach(musicTag -> {
            repos.load(musicTag);
            updatedItems.add(musicTag);
        });
        _editItems.postValue(updatedItems); // This will trigger observers
        rebuildDisplayTag(updatedItems);     // This will also trigger observers
    }
}

File: app/src/main/java/apincer/android/mmate/ui/viewmodel/TagsViewModel.kt
package apincer.android.mmate.ui.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import apincer.music.core.model.Track
import apincer.music.core.repository.TagRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.function.Function
import javax.inject.Inject

@HiltViewModel
class TagsViewModel(
    private val repos: TagRepository,
    private val ioDispatcher: CoroutineDispatcher
) : ViewModel() {

    @Inject
    constructor(repos: TagRepository) : this(repos, Dispatchers.IO)

    companion object {
        const val MULTIPLE_ALBUMS = "[Multiple Albums]"
        const val MULTIPLE_ARTISTS = "[Multiple Artists]"
        const val MULTIPLE_ALBUM_ARTISTS = "[Multiple Album Artists]"
        const val MULTIPLE_EMPTY = ""

        /**
         * Generic method to find a common string value from a list of Tracks for a given field.
         */
        @JvmStatic
        fun getCommonStringValue(
            tags: List<Track>?,
            valueExtractor: Function<Track, String?>,
            multipleValueString: String,
            defaultValueForNullCommon: String
        ): String {
            if (tags.isNullOrEmpty()) {
                return multipleValueString
            }

            val values = tags.filterNotNull().map { valueExtractor.apply(it) }
            if (values.isEmpty()) {
                return multipleValueString
            }

            val distinctCount = values.distinct().take(2).size
            return if (distinctCount > 1) {
                multipleValueString
            } else {
                values[0] ?: defaultValueForNullCommon
            }
        }
    }

    // --- Core Data ---
    private val _editItems = MutableLiveData<List<Track>>(emptyList())
    @JvmField
    val editItems: LiveData<List<Track>> = _editItems

    private val _displayTag = MutableLiveData<Track?>()
    @JvmField
    val displayTag: LiveData<Track?> = _displayTag

    // StateFlow equivalents for modern Compose screens
    private val _editItemsFlow = MutableStateFlow<List<Track>>(emptyList())
    val editItemsFlow: StateFlow<List<Track>> = _editItemsFlow.asStateFlow()

    private val _displayTagFlow = MutableStateFlow<Track?>(null)
    val displayTagFlow: StateFlow<Track?> = _displayTagFlow.asStateFlow()

    private val _studioProvenanceFlow = MutableStateFlow(StudioProvenanceInfo())
    val studioProvenanceFlow: StateFlow<StudioProvenanceInfo> = _studioProvenanceFlow.asStateFlow()

    private val _relatedTracksSheetState = MutableStateFlow(RelatedTracksSheetState())
    val relatedTracksSheetState: StateFlow<RelatedTracksSheetState> = _relatedTracksSheetState.asStateFlow()

    private fun setEditItems(items: List<Track>) {
        _editItemsFlow.value = items
        _editItems.postValue(items)
    }

    private fun setDisplayTag(tag: Track?) {
        _displayTagFlow.value = tag
        _displayTag.postValue(tag)
        loadStudioProvenance(tag)
    }

    fun processAudioTagEditEvent(items: List<Track>?) {
        val safeItems = items ?: emptyList()
        setEditItems(safeItems)
        redisplayTag(safeItems)
    }

    fun updateWithPlayingSong(playingSong: Track?) {
        if (playingSong == null) return
        viewModelScope.launch(ioDispatcher) {
            repos.load(playingSong)
            val singleItemList = listOf(playingSong.copy())
            withContext(Dispatchers.Main) {
                setEditItems(singleItemList)
                redisplayTag(singleItemList)
            }
        }
    }

    fun redisplayTag(currentItems: List<Track>?) {
        if (currentItems.isNullOrEmpty()) {
            setDisplayTag(null)
            return
        }

        val newDisplayTag = currentItems[0].copy()
        if (currentItems.size > 1) {
            newDisplayTag.title = "[${currentItems.size} songs selected]"
            newDisplayTag.album = getCommonStringValue(currentItems, { it.album }, MULTIPLE_ALBUMS, MULTIPLE_EMPTY)
            newDisplayTag.artist = getCommonStringValue(currentItems, { it.artist }, MULTIPLE_ARTISTS, MULTIPLE_EMPTY)
            newDisplayTag.albumArtist = getCommonStringValue(currentItems, { it.albumArtist }, MULTIPLE_ALBUM_ARTISTS, MULTIPLE_EMPTY)
            newDisplayTag.genre = getCommonStringValue(currentItems, { it.genre }, MULTIPLE_EMPTY, MULTIPLE_EMPTY)
            newDisplayTag.track = getCommonStringValue(currentItems, { it.track }, MULTIPLE_EMPTY, MULTIPLE_EMPTY)
            newDisplayTag.year = getCommonStringValue(currentItems, { it.year }, MULTIPLE_EMPTY, MULTIPLE_EMPTY)
            newDisplayTag.publisher = getCommonStringValue(currentItems, { it.publisher }, MULTIPLE_EMPTY, MULTIPLE_EMPTY)
        }
        setDisplayTag(newDisplayTag)
    }

    fun refreshDisplayTag() {
        val items = _editItems.value
        if (items.isNullOrEmpty()) return

        viewModelScope.launch(ioDispatcher) {
            val updatedItems = ArrayList<Track>(items.size)
            for (musicTag in items) {
                repos.load(musicTag)
                updatedItems.add(musicTag)
            }
            withContext(Dispatchers.Main) {
                setEditItems(updatedItems)
                redisplayTag(updatedItems)
            }
        }
    }

    fun loadStudioProvenance(track: Track?) {
        if (track == null) {
            _studioProvenanceFlow.value = StudioProvenanceInfo()
            return
        }

        viewModelScope.launch(ioDispatcher) {
            val artist = track.artist?.trim().orEmpty()
            val album = track.album?.trim().orEmpty()
            val path = track.path?.let { java.io.File(it).parent }.orEmpty()
            val folderName = if (path.isNotEmpty()) java.io.File(path).name else ""

            var artistCount = 0
            var albumCount = 0
            var folderCount = 0

            if (artist.isNotEmpty() && !artist.startsWith("[")) {
                val criteria = apincer.music.core.model.SearchCriteria(apincer.music.core.model.SearchCriteria.TYPE.LIBRARY).apply {
                    filterType = apincer.music.core.Constants.FILTER_TYPE_ARTIST
                    filterText = artist
                }
                artistCount = repos.getSearchStats(criteria)?.totalCount ?: 0
            }

            if (album.isNotEmpty() && !album.startsWith("[")) {
                val criteria = apincer.music.core.model.SearchCriteria(apincer.music.core.model.SearchCriteria.TYPE.LIBRARY).apply {
                    filterType = apincer.music.core.Constants.FILTER_TYPE_ALBUM
                    filterText = album
                }
                albumCount = repos.getSearchStats(criteria)?.totalCount ?: 0
            }

            if (path.isNotEmpty()) {
                val criteria = apincer.music.core.model.SearchCriteria(apincer.music.core.model.SearchCriteria.TYPE.LIBRARY).apply {
                    filterType = apincer.music.core.Constants.FILTER_TYPE_PATH
                    filterText = path
                }
                folderCount = repos.getSearchStats(criteria)?.totalCount ?: 0
            }

            _studioProvenanceFlow.value = StudioProvenanceInfo(
                artist = artist,
                artistCount = artistCount,
                album = album,
                albumCount = albumCount,
                folderName = folderName,
                folderPath = path,
                folderCount = folderCount
            )
        }
    }

    fun openRelatedTracks(filterType: String, filterKeyword: String, title: String) {
        if (filterKeyword.isBlank()) return
        _relatedTracksSheetState.value = RelatedTracksSheetState(
            isVisible = true,
            title = title,
            subtitle = "Loading studio tracks...",
            filterType = filterType,
            filterKeyword = filterKeyword,
            tracks = emptyList(),
            isLoading = true
        )

        viewModelScope.launch(ioDispatcher) {
            val criteria = apincer.music.core.model.SearchCriteria(apincer.music.core.model.SearchCriteria.TYPE.LIBRARY).apply {
                this.filterType = filterType
                this.filterText = filterKeyword
            }
            val results = repos.findMusic(criteria) ?: emptyList()
            val stats = repos.getSearchStats(criteria)
            val count = (stats?.totalCount ?: results.size.toLong()).toInt()
            val durationMin = if (stats != null && stats.totalDuration > 0) {
                val totalSec = stats.totalDuration.toLong()
                val hrs = totalSec / 3600
                val mins = (totalSec % 3600) / 60
                if (hrs > 0) "$hrs hr $mins min" else "$mins min"
            } else ""
            val subtitle = if (durationMin.isNotEmpty()) "$count Studio Tracks • $durationMin" else "$count Studio Tracks"

            _relatedTracksSheetState.value = RelatedTracksSheetState(
                isVisible = true,
                title = title,
                subtitle = subtitle,
                filterType = filterType,
                filterKeyword = filterKeyword,
                tracks = results,
                isLoading = false
            )
        }
    }

    fun closeRelatedTracks() {
        _relatedTracksSheetState.value = _relatedTracksSheetState.value.copy(isVisible = false)
    }
}

data class StudioProvenanceInfo(
    val artist: String = "",
    val artistCount: Int = 0,
    val album: String = "",
    val albumCount: Int = 0,
    val folderName: String = "",
    val folderPath: String = "",
    val folderCount: Int = 0
)

data class RelatedTracksSheetState(
    val isVisible: Boolean = false,
    val title: String = "",
    val subtitle: String = "",
    val filterType: String = "",
    val filterKeyword: String = "",
    val tracks: List<Track> = emptyList(),
    val isLoading: Boolean = false
)
